Expand on the fallback context in web integration (#119)
tc39/proposal-async-context: WEB-INTEGRATION.md

        }
        this.dispatchEvent(new Event("load"));
      } catch (e) {
        this.dispatchEvent(new Event("error"));
      }
    })();
  }
}
```

And lets trace how the context propagates from `.send()` in the following case:
```javascript
const asyncVar = new AsyncContext.Variable();
const xhr = new XMLHttpRequest();
xhr.open("GET", "https://example.com");
xhr.addEventListener("load", () => {
  console.log(asyncVar.get()); // "foo"
});
asyncVar.run("foo", () => {
  xhr.send();
});
```
- when `.send()` is called, the value of `asyncVar` is `"foo"`.
- it is synchronously propagated up to the `fetch()` call in `.send()`
- the `await` snapshots the context before pausing, and restores it (to `asyncVar: "foo"`) when the `fetch` completes
- the `await`s in the reader loop propagate the context as well
- when `this.dispatchEvent(new Event("load"))`, is called, the current active async context is thus
  the same one as when `.send()` was called
- the `"load"` callback thus runs with `asyncVar` set to `"foo"`.

Note that this example uses `await`, but due to the proposed semantics for `.then` and `setTimeout`

> (and similar APIs), the same would hapepn when using other asynchronicity primitives.

Reviewer: Maybe add something like "If something isn't expressible using existing asynchronicity primitives, we assume that any internal primitives would take callbacks that would propagate the context as described above."

Things like "queue a task" are similar but not identical to `setTimeout(cb, 0)`, and if queuing a task in response to e.g. a websocket request, that couldn't exactly be expressed in terms of anything already in the language.

Author: To avoid "this is based on something that is not actually possible", I'll mention that you can still _emulate_ those API (e.g. you can mock the WebSocket API so that it looks like you are dealing with a real web socket, but you are actually not).

</details>

> Event listeners for events dispatched **from the browser** rather than as a consequence of some JS action (e.g. a user clicking on a button) will by default run in the root (empty) context. This is the same
> context that the browser uses, for example, for the top-level execution of scripts.

Reviewer: ```suggestion
Event listeners for events dispatched **from the browser** rather than as a consequence of some JS action (e.g. a user clicking on a button) will by default run in the root (empty) context. This is the same
context that the browser uses, for example, for the top-level execution of scripts.
This also applies to events originating from a different agent or cross-origin window.
```

Author: The next line says that already.

    <button id="prev">Previous day</button>
    <output>...</output>
    <button id="next">Next day</button>
  `;
  document.body.appendChild(widget);

  const load = async () => {
    const response = await fetch(`/weather/${day}`);
    widget.querySelector("output").textContent = await response.text();
  };

  widget.querySelector("#prev").addEventListener("click", async () => {
    day = day.subtract({ days: 1 });
    await load();
  });
  widget.querySelector("#next").addEventListener("click", async () => {
    day = day.add({ days: 1 });
    await load();
  });

  load();
}
```

When the user clicks on one of the buttons and the `fetch` it triggers fails,
without using `captureFallbackContext` the `unhandledrejection` event listener
would not know that the failure is coming from the `weather-widget` widget.

Thanks to `captureFallbackContext`, that information is properly propagated.

</details>

This fallback is per-variable and not based on `AsyncContext.Snapshot`, to avoid
accidentally keeping alive unnecessary objects.

There are still some questions about `captureFallbackContext` that need to be
answered:
- should it take just one variable or a list of variables?

> - should it just be for event targets, or for all web APIs that can be triggered
>   by non-JS code?

Reviewer: ```suggestion
- should it just be for event targets, or also for web APIs that take a callback
  which can run when triggered by non-JS code?
```

Author: Changing this to:
> - should it just be for event targets, or for all web APIs that take a callback which can run when triggered not by JavaScript?

I don't like the "or also for web APIs that take a callback which ..." because events are in that category, it's not an "or also".

  xhr.send();
});
```
- when `.send()` is called, the value of `asyncVar` is `"foo"`.
- it is synchronously propagated up to the `fetch()` call in `.send()`
- the `await` snapshots the context before pausing, and restores it (to `asyncVar: "foo"`) when the `fetch` completes
- the `await`s in the reader loop propagate the context as well
- when `this.dispatchEvent(new Event("load"))`, is called, the current active async context is thus
  the same one as when `.send()` was called
- the `"load"` callback thus runs with `asyncVar` set to `"foo"`.

Note that this example uses `await`, but due to the proposed semantics for `.then` and `setTimeout`
(and similar APIs), the same would happen when using other asynchronicity primitives. Note that most APIs
dealing with I/O are not actually polyfillable in JavaScript, but you can still emulate/mock them with
testing data.

</details>

Event listeners for events dispatched **from the browser** rather than as a consequence of some JS action (e.g. a user clicking on a button) will by default run in the root (empty) context. This is the same
context that the browser uses, for example, for the top-level execution of scripts.

> > NOTE: To keep agents isolated, events dispatched from different agents (e.g. from a worker, or from a cross-origin iframe) will behave like events dispatched by user interaction.

Reviewer: ```suggestion
> NOTE: To keep agents isolated, events dispatched from different agents (e.g. from a worker, or from a cross-origin iframe) will behave like events dispatched by user interaction. This also applies to events dispatched from cross-origin iframes in the same agent, to avoid exposing the fact that they're in the same agent.
```
